src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala

```diff
@@ -568,4 +588,56 @@ object RuleUtils {
     assert(shuffleInjected)
     shuffled
   }
+
+  private def transformProject(project: Project, index: IndexLogEntry): Project = {
+    val projectedFields = project.projectList.map { exp =>
+      val fieldName = extractNamesFromExpression(exp).head
+      val escapedFieldName = SchemaUtils.prefixNestedFieldName(fieldName)
+      val attr = extractAttributeRef(exp, fieldName)
+      val fieldType = extractTypeFromExpression(exp, fieldName)
+      val exprId = getExprId(project, escapedFieldName)
```

Comment:
Does this work as expected? I think it's because of 'transformDown', transformProject will be called before transforming the relation node. 
Could you add some test for filter index rule?

Reply:
@sezruby: This functionality is already tested by existing UTs and E2Es. 

This is the correct approach: **Setting the `exprId` on the lowest leaf (aka: relation output) and when transforming each node get that expression id and use it when transforming that node**. So, the same expression id used in the relation is used in the tree otherwise there will be thrown analysis exception saying that the field (aka attribute) could not be found. I cannot set the `exprId` with a value on the relation, then in the filter and projection use other expression id (if I would use `newExprId` again). I have to take it from the relation that I previously created.

For example, given the `__hs_nested.nest.leaf.field`, in the relation I set it to a new number given by `newExprId` (ie: `#701`) and then on the filter to get another `exprId` (ie: `#801`), although at planning it will not crash it will throw exception at runtime.

This is how the correct plan should look like:

```
Project [__hs_nested.nested.nest.field#701]
+- Filter (isnotnull(__hs_nested.nested.nest.field#701) && (__hs_nested.nested.nest.field#701 = wa1))
   +- Relation[__hs_nested.nested.nest.field#701]
```

As you can see that `#701` is everywhere.

Now, regarding the `transformDown` and why it works, I guess the answer is that it is running multiple times.

I'll try switching to `transformUp` and see how it goes.